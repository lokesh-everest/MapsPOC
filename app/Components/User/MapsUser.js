import React from 'react'
import {Button, View, Image, StyleSheet, Dimensions,Text} from 'react-native';
import MapView, {Marker} from "react-native-maps";
import MapViewDirections from 'react-native-maps-directions';
import Config from 'react-native-config';
import Statistics from "../Statistics";

export default class MapsUser extends React.Component {
    constructor(props) {
        super(props);
        this.mapRef = null;
        this.state = {
            distance: 0,
            duration: 0,
            sourceCoordinates: this.props.sourceCoordinates,
            destinationCoordinates: this.props.destinationCoordinates,
            driverCoordinates: this.props.sourceCoordinates,
            initialMap: this.props.initialMap
        };
        this.fitToMarkers = this.fitToMarkers.bind(this);
    }

    componentWillReceiveProps(nextProps) {
        this.setState({driverCoordinates: nextProps.driverCoordinates});
    }

    fitToMarkers() {
        this.mapRef.fitToCoordinates([this.state.driverCoordinates, this.state.destinationCoordinates], {
            edgePadding: {top: 50, right: 20, bottom: 20, left: 20},
            animated: true
        })
    }

    moveDriverSmoothly(updatedCoords, timeToTraverse) {
        if (this.marker) {
            this.marker._component.animateMarkerToCoordinate(updatedCoords, timeToTraverse);
        }
    }

    _gotoCurrentLocation() {
        this.mapRef.animateToRegion({
            latitude: this.state.driverCoordinates.latitude,
            longitude: this.state.driverCoordinates.longitude,
            latitudeDelta: 0.0059397161733585335,
            longitudeDelta: 0.005845874547958374
        });
    }

    render() {
        return (
            <View style={{flex: 1}}>
                <Text>User</Text>
                <MapView
                    ref={(ref) => {
                        this.mapRef = ref
                    }}
                    style={{flex: 1}}
                    initialRegion={this.state.initialMap}
                    showsUserLocation={false}
                    onMapReady={this.fitToMarkers}>
                    {
                        this.props.markers.map(marker => (
                            <Marker key={marker.key} coordinate={marker.coordinates} title={marker.title}>
                                <Image style={{width: 30, height: 30}} source={marker.imagePath}/>
                            </Marker>
                        ))
                    }
                    <MapViewDirections
                        origin={this.state.driverCoordinates}
                        destination={this.state.destinationCoordinates}
                        apikey={Config.GOOGLE_MAPS_API_KEY}
                        waypoints={this.props.waypoints}
                        strokeColor={"#0F85BF"}
                        strokeWidth={7}
                        resetOnChange={false}
                        onReady={result => {
                            const duration = Math.ceil(result.duration);
                            const distance = Math.round(result.distance * 100) / 100;
                            let timeToTraverse = 2000;
                            const distanceMoved = Math.abs(this.state.distance - distance);
                            if (distanceMoved < 0.5) {
                                timeToTraverse = (distanceMoved / 0.2) * 1000;
                            }
                            this.moveDriverSmoothly(this.state.driverCoordinates, timeToTraverse);
                            distance > 1 ? this._gotoCurrentLocation() : this.fitToMarkers();
                            this.setState({duration: duration, distance: distance})
                        }}
                    />
                    <Marker.Animated
                        ref={marker => {
                        this.marker = marker;}}
                        coordinate={this.state.driverCoordinates} title={"Driver"}>
                        <Image style={{width: 30, height: 30}} source={require("./../../assets/delievery.png")}/>
                    </Marker.Animated>
                </MapView>
                <View style={styles.text}>
                    <Statistics duration={this.state.duration} distance={this.state.distance}/>
                </View>
                <Button title='Fit' onPress={this.fitToMarkers}/>
            </View>
        )
    }
}
const styles=StyleSheet.create({
    text:{
        bottom: 100,
        left: Dimensions.get('window').width/2-35,
        position: "absolute",
        backgroundColor:"black",
    },
    map:{
        flex:1,
        zIndex: -1
    },
    button:{
        bottom:100,
        left:Dimensions.get('window').width/2-90,
        position:"absolute",
    }
})