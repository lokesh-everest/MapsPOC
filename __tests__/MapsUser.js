import 'jsdom-global/register';
import 'react-native';
import React from 'react';
import MapsUser from "../app/Components/User/MapsUser";

import renderer from 'react-test-renderer';
import { shallow, mount, configure } from 'enzyme';

describe(MapsUser, () => {
    it('render correcly', () => {
        const wrapper = shallow(<MapsUser markers={[{ 'latitude': 0, 'longitude': 0 }]} />);
        expect(wrapper).toMatchSnapshot();
    })
})
