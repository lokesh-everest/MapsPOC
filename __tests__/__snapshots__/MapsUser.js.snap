// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapsUser render correcly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text>
    User
  </Text>
  <MapView
    onMapReady={[Function]}
    showsUserLocation={true}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <MapMarker
      stopPropagation={false}
    >
      <Component
        style={
          Object {
            "height": 30,
            "width": 30,
          }
        }
      />
    </MapMarker>
    <MapViewDirections
      onReady={[Function]}
      strokeColor="blue"
      strokeWidth={3}
    />
    <MapMarker
      stopPropagation={false}
      title="Driver"
    >
      <Component
        source={
          Object {
            "testUri": "../../../app/assets/delievery.png",
          }
        }
        style={
          Object {
            "height": 30,
            "width": 30,
          }
        }
      />
    </MapMarker>
  </MapView>
  <View
    style={
      Object {
        "backgroundColor": "black",
        "bottom": 100,
        "left": 340,
        "position": "absolute",
      }
    }
  >
    <Statitics
      distance={0}
      duration={0}
    />
  </View>
</View>
`;
